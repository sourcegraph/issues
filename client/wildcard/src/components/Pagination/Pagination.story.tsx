import { storiesOf } from '@storybook/react'
import React, { useState } from 'react'
import { number } from '@storybook/addon-knobs'
import webStyles from '../../../../web/src/SourcegraphWebApp.scss'
import { Pagination } from './Pagination'

const { add } = storiesOf('wildcard/Pagination', module).addDecorator(story => (
    <>
        <style>{webStyles}</style>
        <div className="layout__app-router-container">
            <div className="container web-content mt-3">{story()}</div>
        </div>
    </>
))

add('Basic', () => {
    const [page, setPage] = useState(1)
    return <Pagination currentPage={page} onPageChange={setPage} maxPages={number('maxPages', 10)} />
})
