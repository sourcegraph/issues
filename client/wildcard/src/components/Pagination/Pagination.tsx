import React from 'react'
import classNames from 'classnames'
import ChevronLeftIcon from 'mdi-react/ChevronLeftIcon'
import ChevronRightIcon from 'mdi-react/ChevronRightIcon'
import useResizeObserver from 'use-resize-observer/polyfilled'
import { useDebounce } from '../../hooks/useDebounce'
import { omit } from 'lodash'
import { usePagination } from '../../hooks/usePagination'

interface PageButtonProps extends React.ButtonHTMLAttributes<HTMLButtonElement> {
    active?: boolean
}

const PageButton = React.forwardRef<HTMLButtonElement, PageButtonProps>(({ children, active, ...props }, reference) => (
    <button
        type="button"
        className={classNames('btn', 'mx-1', 'pagination__button', active && 'btn-primary')}
        aria-current={active}
        ref={reference}
        {...props}
    >
        {children}
    </button>
))

export interface PaginationProps {
    currentPage: number
    onPageChange: (page: number) => void
    totalPages: number
    className?: string
}

const validatePositiveInteger = (name: string, value: number): string[] => {
    if (value > 0) {
        return []
    }
    return [`${name} must have a value greater than 0`]
}

const validateProps = (props: PaginationProps): void => {
    const errors = [
        ...validatePositiveInteger('totalPages', props.totalPages),
        ...validatePositiveInteger('currentPage', props.currentPage),
    ]

    if (props.currentPage > props.totalPages) {
        errors.push('currentPage must be not be greater than totalPages')
    }

    if (errors.length > 0) {
        throw new Error(`Invalid configuration:\n${errors.map(error => `â€” ${error}`).join('\n')}`)
    }
}

export const Pagination: React.FunctionComponent<PaginationProps> = props => {
    validateProps(props)

    const { totalPages, currentPage, className, onPageChange } = props
    const { ref, width } = useDebounce(useResizeObserver(), 100)
    const shouldShrink = width !== undefined && width <= 576
    const pages = usePagination({
        currentPage,
        onChange: onPageChange,
        totalPages,
        maxDisplayed: shouldShrink ? 3 : 5,
    })

    return (
        <nav>
            <ul ref={ref} className={classNames('pagination', className)}>
                {pages.map((page, index) => {
                    if (page.type === 'previous' || page.type === 'next') {
                        return (
                            <li key={index}>
                                <PageButton {...omit(page, 'type')}>
                                    {page.type === 'previous' && (
                                        <ChevronLeftIcon className="icon-inline" aria-hidden={true} />
                                    )}
                                    <span className={classNames(shouldShrink && 'd-none')}>{page.content}</span>
                                    {page.type === 'next' && (
                                        <ChevronRightIcon className="icon-inline" aria-hidden={true} />
                                    )}
                                </PageButton>
                            </li>
                        )
                    }

                    return (
                        <li key={index}>
                            <PageButton {...omit(page, 'type')}>{page.content}</PageButton>
                        </li>
                    )
                })}
            </ul>
        </nav>
    )
}
