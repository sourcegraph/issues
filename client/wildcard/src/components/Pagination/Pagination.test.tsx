import React from 'react'
import { render, RenderResult, cleanup, fireEvent } from '@testing-library/react'
import { Pagination, Props } from './Pagination'
import sinon from 'sinon'

describe('Pagination', () => {
    let queries: RenderResult
    const renderWithProps = (props: Props): RenderResult => render(<Pagination {...props} />)
    const onPageChangeMock = sinon.spy(() => undefined)

    beforeEach(() => {
        onPageChangeMock.resetHistory()
    })

    afterEach(cleanup)

    describe('Invalid configuration', () => {
        it('will error when less than 1 max pages', () => {
            expect(() => {
                renderWithProps({ currentPage: 0, maxPages: 0, onPageChange: onPageChangeMock })
            }).toThrowErrorMatchingSnapshot()
        })

        it('will error when currentPage is less than 1', () => {
            expect(() => {
                renderWithProps({ currentPage: -1, maxPages: 10, onPageChange: onPageChangeMock })
            }).toThrowErrorMatchingSnapshot()
        })

        it('will error when currentPage is greater than maxPages', () => {
            expect(() => {
                renderWithProps({ currentPage: 11, maxPages: 10, onPageChange: onPageChangeMock })
            }).toThrowErrorMatchingSnapshot()
        })
    })

    describe('Typical pagination', () => {
        let currentPage = 2
        beforeEach(() => {
            queries = renderWithProps({ currentPage, maxPages: 10, onPageChange: onPageChangeMock })
        })

        it('will render correct elipsis', () => {
            expect(queries.getAllByRole('button', { name: '...' })).toHaveLength(1)
        })

        it('will render correct pages', () => {
            const expectedPages = ['1', '2', '3', '4', '5', '10']
            for (const page of expectedPages) {
                expect(queries.getByRole('button', { name: page })).toBeInTheDocument()
            }
        })

        it('will render correct currently selected page', () => {
            expect(queries.getByRole('button', { name: String(currentPage) })).toHaveAttribute('aria-current', 'true')
        })

        it('will render previous and next buttons', () => {
            expect(queries.getByText('Previous')).toBeInTheDocument()
            expect(queries.getByText('Next')).toBeInTheDocument()
        })

        it('calls onPageChange correctly when page is individually selected', () => {
            fireEvent.click(queries.getByRole('button', { name: '5' }))
            expect(onPageChangeMock).toHaveBeenCalledTimes(1)
            expect(onPageChangeMock).toHaveBeenCalledWith(5)
        })

        it('calls onPageChange correctly when previous button is selected', () => {
            fireEvent.click(queries.getByText('Previous'))
            expect(onPageChangeMock).toHaveBeenCalledTimes(1)
            expect(onPageChangeMock).toHaveBeenCalledWith(currentPage - 1)
        })

        it('calls onPageChange correctly when next button is selected', () => {
            fireEvent.click(queries.getByText('Next'))
            expect(onPageChangeMock).toHaveBeenCalledTimes(1)
            expect(onPageChangeMock).toHaveBeenCalledWith(currentPage + 1)
        })

        describe('when currently selected near middle', () => {
            beforeAll(() => {
                currentPage = 5
            })

            it('will render correct elipsis', () => {
                expect(queries.getAllByRole('button', { name: '...' })).toHaveLength(2)
            })

            it('will render correct pages', () => {
                const expectedPages = ['1', '4', '5', '6', '10']
                for (const page of expectedPages) {
                    expect(queries.getByRole('button', { name: page })).toBeInTheDocument()
                }
            })
        })

        describe('when currently selected near end', () => {
            beforeAll(() => {
                currentPage = 10
            })

            it('will render correct elipsis', () => {
                expect(queries.getAllByRole('button', { name: '...' })).toHaveLength(1)
            })

            it('will render correct pages', () => {
                const expectedPages = ['1', '6', '7', '8', '9', '10']
                for (const page of expectedPages) {
                    expect(queries.getByRole('button', { name: page })).toBeInTheDocument()
                }
            })
        })
    })

    describe('Small pagination', () => {
        const currentPage = 2
        beforeEach(() => {
            queries = renderWithProps({ currentPage, maxPages: 3, onPageChange: onPageChangeMock })
        })

        it('will render no elipsis', () => {
            expect(queries.queryAllByRole('button', { name: '...' })).toHaveLength(0)
        })

        it('will render correct pages', () => {
            const expectedPages = ['1', '2', '3']
            for (const page of expectedPages) {
                expect(queries.getByRole('button', { name: page })).toBeInTheDocument()
            }
        })

        it('will render correct currently selected page', () => {
            expect(queries.getByRole('button', { name: String(currentPage) })).toHaveAttribute('aria-current', 'true')
        })
    })

    describe('Large pagination', () => {
        const currentPage = 1
        beforeEach(() => {
            queries = renderWithProps({ currentPage, maxPages: 30, onPageChange: onPageChangeMock })
        })

        it('will render correct elipsis', () => {
            expect(queries.getAllByRole('button', { name: '...' })).toHaveLength(1)
        })

        it('will render correct pages', () => {
            const expectedPages = ['1', '2', '3', '4', '5', '30']
            for (const page of expectedPages) {
                expect(queries.getByRole('button', { name: page })).toBeInTheDocument()
            }
        })

        it('will render correct currently selected page', () => {
            expect(queries.getByRole('button', { name: String(currentPage) })).toHaveAttribute('aria-current', 'true')
        })
    })
})
