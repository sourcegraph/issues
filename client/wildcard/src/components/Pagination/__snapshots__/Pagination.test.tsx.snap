// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pagination Invalid configuration will error when currentPage is greater than totalPages 1`] = `
"Invalid configuration:
— currentPage must be not be greater than totalPages"
`;

exports[`Pagination Invalid configuration will error when currentPage is less than 1 1`] = `
"Invalid configuration:
— currentPage must have a value greater than 0"
`;

exports[`Pagination Invalid configuration will error when less than 1 max pages 1`] = `
"Invalid configuration:
— totalPages must have a value greater than 0
— currentPage must have a value greater than 0"
`;
